cpp/src/arrow/dataset/discovery_test.cc

```diff
@@ -31,30 +31,34 @@ class FileSystemDataSourceDiscoveryTest : public TestFileSystemBasedDataSource {
  public:
   void MakeDiscovery(const std::vector<fs::FileStats>& files) {
     MakeFileSystem(files);
-    ASSERT_OK(
-        FileSystemDataSourceDiscovery::Make(fs_.get(), selector_, format_, &discovery_));
+    ASSERT_OK(FileSystemDataSourceDiscovery::Make(fs_.get(), selector_, format_,
+                                                  discovery_options_, &discovery_));
+  }
+
+  void AssertFinishWithPaths(std::vector<std::string> paths) {
+    ASSERT_OK(discovery_->Finish(&source_));
+    AssertFragmentsAreFromPath(source_->GetFragments(options_), paths);
   }
 
  protected:
   fs::Selector selector_;
+  FileSystemDiscoveryOptions discovery_options_;
   std::shared_ptr<DataSourceDiscovery> discovery_;
   std::shared_ptr<FileFormat> format_ = std::make_shared<DummyFileFormat>();
 };
 
 TEST_F(FileSystemDataSourceDiscoveryTest, Basic) {
   MakeDiscovery({fs::File("a"), fs::File("b")});
 
-  ASSERT_OK(discovery_->Finish(&source_));
-  AssertFragmentsAreFromPath(source_->GetFragments(options_), {"a", "b"});
+  AssertFinishWithPaths({"a", "b"});
 }
 
 TEST_F(FileSystemDataSourceDiscoveryTest, Selector) {
   selector_.base_dir = "A";
   MakeDiscovery({fs::File("0"), fs::File("A/a")});
 
-  ASSERT_OK(discovery_->Finish(&source_));
   // "0" doesn't match selector, so it has been dropped:
-  AssertFragmentsAreFromPath(source_->GetFragments(options_), {"A/a"});
+  AssertFinishWithPaths({"A/a"});
```

Comment:
Please also add a test with `partition_base_dir` which is distinct from selector base_dir